ejb3/src/main/java/org/jboss/as/ejb3/remote/AssociationImpl.java

```diff
@@ -73,23 +73,59 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.CancellationException;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
 import java.util.concurrent.Future;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
 
 /**
  * @author <a href="mailto:tadamski@redhat.com">Tomasz Adamski</a>
  */
-final class AssociationImpl implements Association {
+final class AssociationImpl implements Association, Registry.Listener<String, List<ClientMapping>>, AutoCloseable {
```

Comment:
I don't think that the fact that events are queued sequentially and the fact that they are per-connection are logically connected in any way (each connection independently has a sequential view, and each connection is separately asynchronous in terms of when it receives/processes events), thus from that perspective it doesn't seem to matter whether there is a listener per connection or a single listener which is delegated to many connections.

Also it's worth noting that there is exactly one consumer of the Registry API, so it's reasonable to expect that the API matches the needs of the consumer as closely as possible, as opposed to requiring a second layer of listener tracking.